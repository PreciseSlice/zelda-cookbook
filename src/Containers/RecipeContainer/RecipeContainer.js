import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import RecipeCard from '../../Components/RecipeCard/RecipeCard';
import './RecipeContainer.css';

export class RecipeContainer extends Component {
  constructor(props) {
    super(props);

    this.state = {
      loading: true,
    };
  }

  filterByIngredient = recipes => {    
    const { ingredientFilter } = this.props;

    if (ingredientFilter.length) {
      const filtered = recipes.filter( recipe => {
        let match = 0;
        ingredientFilter.forEach( ingredient => {
          for (let i = 1; i <= 5; i++) {
            if (parseInt(ingredient.id, 10) === recipe[`ingredient${i}_id`]) {
              match++;
            }
          }
        });
        return match === ingredientFilter.length;
      });
      return filtered;
    } else {
      return recipes;
    }
  }

  filterByName = recipes => {
    const { nameFilter } = this.props;
    if (nameFilter) {
      return recipes.filter(recipe => recipe.name.toLowerCase().includes(
        nameFilter.toLowerCase())
      );
    } else {
      return recipes;
    }
  }

  filterByType = recipes => {
    const { typeFilter } = this.props;
    if (typeFilter) {
      return recipes.filter(recipe => recipe.type === typeFilter);
    } else {
      return recipes;
    }
  }

  filterRecipes = (allRecipes) => {
    const onceFiltered = this.filterByIngredient(allRecipes);
    const twiceFiltered = this.filterByName(onceFiltered);
    const thriceFiltered = this.filterByType(twiceFiltered);
    return thriceFiltered;
  }

  renderRecipes = () => {
    const recipes = this.filterRecipes(this.props.recipes);

    if(recipes.length) {      
      const recipesToRender = recipes.map(recipe => {
        return <RecipeCard
          key={ recipe.id }
          id={ recipe.id }
          image={ recipe.image }
          name={ recipe.name }
          hearts={ recipe.hearts }
          type={ recipe.type }
          duration={ recipe.duration }
          notes={ recipe.notes }
          ingredient1={recipe.ingredient1}
          ingredient2={ recipe.ingredient2 }
          ingredient3={ recipe.ingredient3 }
          ingredient4={ recipe.ingredient4 } 
          ingredient5={ recipe.ingredient5 }
          strength={ recipe.strength }
          resale={ recipe.resale }
        />;
      });
      return recipesToRender;
    } else {
    return <div id="loading"></div>;
    }
  }

  render() {
    return (
      <div id="all-recipe-container">
        { this.renderRecipes() }
      </div>
    );
  }
}

const { arrayOf, shape, number, string, array } = PropTypes;

RecipeContainer.propTypes = {
  recipes: arrayOf(shape({
    id: number,
    category: string,
    hearts: string,
    name: string,
    notes: string,
    resale: string,
    type: string,
    duration: string,
    strength: string,
    ingredient1: string,
    ingredient2: string,
    ingredient3: string,
    ingredient4: string,
    ingredient5: string,
    ingredient1_id: number,
    ingredient2_id: number,
    ingredient3_id: number,
    ingredient4_id: number,
    ingredient5_id: number,
    created_at: string,
    updated_at: string
  })),
  ingredientFilter: array,
  nameFilter: string,
  typeFilter: string
};

export const MSTP = (props) => {
  const {
    recipes,
    ingredientFilter,
    nameFilter,
    typeFilter
  } = props;

  return {
    recipes,
    ingredientFilter,
    nameFilter,
    typeFilter
  };
};

export default connect(MSTP)(RecipeContainer);
