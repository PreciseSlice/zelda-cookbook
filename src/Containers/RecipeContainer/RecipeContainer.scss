#loading {
  background-image: url('https://68.media.tumblr.com/7a5f3925e8e8351619759a2a9566bb03/tumblr_omsjs7eiVk1r85hlio1_r1_500.gif');
  background-repeat: no-repeat;
  height: 100%;
  width: 100%;
  margin: auto 25%;
}