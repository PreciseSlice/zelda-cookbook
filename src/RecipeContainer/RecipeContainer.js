import React, { Component } from 'react';
import { getRecipes } from '../apiCalls';
// import { getRecipeData } from '../dataCleaner';
import RecipeCard from '../RecipeCard/RecipeCard';
import './RecipeContainer.css';

export default class RecipeContainer extends Component {
  constructor() {
    super()

    this.state = {
      loading: true,
      rawRecipes: []
    }
  }

  async componentDidMount() {
    try {
      // const recipeType = this.props.location.pathname.slice(1);
      // console.log('recipeType', recipeType)
      const rawRecipes = await getRecipes();
      // const recipes = getRecipeData(rawRecipes);
      // console.log('recipes', recipes)
      this.setState({ rawRecipes })
    } catch(err) {
      console.log(`ERROR ComponentDidMount RecipeContainer: ${err}`);
    }
  }

  renderRecipes = () => {
    const recipes = this.state.rawRecipes

    if(recipes.length > 0) {
      const recipesToRender = recipes.map(recipe => {
        return <RecipeCard key={ recipe.id }
                           id={ recipe.id }
                           name={ recipe.name }
                           hearts={ recipe.hearts }
                           type={ recipe.type }
                           duration={ recipe.duration }
                           notes={ recipe.notes }
                           ingredient1={ recipe.ingredient1 }
                           ingredient2={ recipe.ingredient2 }
                           ingredient3={ recipe.ingredient3 }
                           ingredient4={ recipe.ingredient4 }
                           ingredient5={ recipe.ingredient5 }
                           strength={ recipe.strength }
                           resale={ recipe.resale } />
      })
      return recipesToRender
    }
  }

  render() {
    return (
      <div id="all-recipe-container">
        { this.renderRecipes() }
      </div>
    )
  }
}
