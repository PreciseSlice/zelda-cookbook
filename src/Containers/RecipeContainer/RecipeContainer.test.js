/* eslint-disable */
import React from 'react';
import { shallow } from 'enzyme';
import { RecipeContainer } from './RecipeContainer';

describe('RecipeContainer', () => {
  let wrapper;
  let mockIngredientFilter = []
  let mockNameFilter = ''
  let mockRecipes = [
    {
      id: 1,
      name: 'Baked Apple',
      ingredient1: 'Apple',
      ingredient2: null,
      ingredient3: null,
      ingredient4: null,
      ingredient5: null,
      ingredient1_id: 2,
      ingredient2_id: null,
      ingredient3_id: null,
      ingredient4_id: null,
      ingredient5_id: null,
    },
    {
      id: 2, 
      name: 'Apple Pie',
      ingredient1: 'Apple',
      ingredient2: 'Tabantha Wheat',
      ingredient3: 'Cane Sugar',
      ingredient4: 'Goat Butter',
      ingredient5: null,
      ingredient1_id: 2,
      ingredient2_id: 3,
      ingredient3_id: 4,
      ingredient4_id: 5,
      ingredient5_id: null,
    }
  ]

  beforeEach(() => {
    wrapper = shallow(<RecipeContainer
      ingredientFilter={mockIngredientFilter}
      nameFilter={mockNameFilter}
      recipes={mockRecipes}
    />);
  });

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  describe('filterByIngredient', () => {
    it('should return all recipes if there are no ingredients in the filter', () => {
      expect(
        wrapper
        .instance()
        .filterByIngredient(mockRecipes)
      ).toEqual(mockRecipes)
    })

    it('should return an array of recipes whose ingredients match what is in the filter', () => {
      const wrapper = shallow(<RecipeContainer
        ingredientFilter={[{name: 'Tabantha Wheat', id: '3'}]}
        nameFilter={mockNameFilter}
        recipes={mockRecipes}
      />);
      
      expect(
        wrapper
        .instance()
        .filterByIngredient(mockRecipes)
      ).toEqual([{
        id: 2,
        name: 'Apple Pie',
        ingredient1: 'Apple',
        ingredient2: 'Tabantha Wheat',
        ingredient3: 'Cane Sugar',
        ingredient4: 'Goat Butter',
        ingredient5: null,
        ingredient1_id: 2,
        ingredient2_id: 3,
        ingredient3_id: 4,
        ingredient4_id: 5,
        ingredient5_id: null,
      }])
    })
  })

  describe('filterByName', () => {
    it('should return all ingredients if there is no name in filter', () => {
      expect(
        wrapper
        .instance()
        .filterByName(mockRecipes)
      ).toEqual(mockRecipes)
    })

    it('should return an array with ingredients matching the name filter', () => {
      const wrapper = shallow(<RecipeContainer
        ingredientFilter={mockIngredientFilter}
        nameFilter={'Pie'}
        recipes={mockRecipes}
      />);

      expect(
        wrapper
        .instance()
        .filterByName(mockRecipes)
      ).toEqual([{
        id: 2,
        name: 'Apple Pie',
        ingredient1: 'Apple',
        ingredient2: 'Tabantha Wheat',
        ingredient3: 'Cane Sugar',
        ingredient4: 'Goat Butter',
        ingredient5: null,
        ingredient1_id: 2,
        ingredient2_id: 3,
        ingredient3_id: 4,
        ingredient4_id: 5,
        ingredient5_id: null,
      }])
    })

    it('should not be case sensative', () => {
      const wrapper = shallow(<RecipeContainer
        ingredientFilter={mockIngredientFilter}
        nameFilter={'pIE'}
        recipes={mockRecipes}
      />);

      expect(
        wrapper
        .instance()
        .filterByName(mockRecipes)
      ).toEqual([{
        id: 2,
        name: 'Apple Pie',
        ingredient1: 'Apple',
        ingredient2: 'Tabantha Wheat',
        ingredient3: 'Cane Sugar',
        ingredient4: 'Goat Butter',
        ingredient5: null,
        ingredient1_id: 2,
        ingredient2_id: 3,
        ingredient3_id: 4,
        ingredient4_id: 5,
        ingredient5_id: null,
      }])
    })
  })

  describe('filterByType', () => {
    it('should return all ingredients if there is no filter applied', () => {
      expect(wrapper.instance().filterByType(mockRecipes)).toEqual(mockRecipes);
    })
  })
});