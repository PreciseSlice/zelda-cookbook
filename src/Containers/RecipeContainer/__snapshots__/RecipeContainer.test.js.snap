// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecipeContainer renderRecipes should match the snapshot if there are no recipes to render 1`] = `
<div
  id="loading"
/>
`;

exports[`RecipeContainer renderRecipes should match the snapshot when it renders ingredients 1`] = `
Array [
  <RecipeCard
    duration={undefined}
    hearts={undefined}
    id={1}
    image={undefined}
    ingredient1="Apple"
    ingredient2={null}
    ingredient3={null}
    ingredient4={null}
    ingredient5={null}
    name="Baked Apple"
    notes={undefined}
    resale={undefined}
    strength={undefined}
    type="none"
/>,
  <RecipeCard
    duration={undefined}
    hearts={undefined}
    id={2}
    image={undefined}
    ingredient1="Apple"
    ingredient2="Tabantha Wheat"
    ingredient3="Cane Sugar"
    ingredient4="Goat Butter"
    ingredient5={null}
    name="Apple Pie"
    notes={undefined}
    resale={undefined}
    strength={undefined}
    type="Restore Hearts"
/>,
]
`;

exports[`RecipeContainer should match the snapshot 1`] = `
<div
  id="all-recipe-container"
>
  <RecipeCard
    id={1}
    ingredient1="Apple"
    ingredient2={null}
    ingredient3={null}
    ingredient4={null}
    ingredient5={null}
    key="1"
    name="Baked Apple"
    type="none"
  />
  <RecipeCard
    id={2}
    ingredient1="Apple"
    ingredient2="Tabantha Wheat"
    ingredient3="Cane Sugar"
    ingredient4="Goat Butter"
    ingredient5={null}
    key="2"
    name="Apple Pie"
    type="Restore Hearts"
  />
</div>
`;
